hoot-core/src/main/cpp/hoot/core/geometry/GeometryMerger.cpp

```diff
@@ -0,0 +1,209 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * --------------------------------------------------------------------
+ *
+ * The following copyright notices are generated automatically. If you
+ * have a new notice to add, please use the format:
+ * " * @copyright Copyright ..."
+ * This will properly maintain the copyright information. DigitalGlobe
+ * copyrights will be updated automatically.
+ *
+ * @copyright Copyright (C) 2020 DigitalGlobe (http://www.digitalglobe.com/)
+ */
+
+#include "GeometryMerger.h"
+
+//  GEOS
+#include <geos/util/IllegalArgumentException.h>
+
+//  Hoot
+#include <hoot/core/geometry/GeometryUtils.h>
+#include <hoot/core/geometry/PolygonCompare.h>
+#include <hoot/core/util/ConfigOptions.h>
+#include <hoot/core/util/Log.h>
+#include <hoot/core/util/StringUtils.h>
+
+//  Qt
+#include <QString>
+#include <QStringList>
+
+//  Standard
+#include <thread>
+
+using namespace geos::geom;
+using namespace std;
+
+namespace hoot
+{
+
+GeometryMerger::GeometryMerger()
+  : _maxThreads(ConfigOptions().getCookieCutterAlphaShapeMaxThreads()),
+    _threadsWorking(true),
+    _finishedThreads(0),
+    _geometryCount(0),
+    _counter(0),
+    _updateInterval(ConfigOptions().getTaskStatusUpdateInterval())
+{
+}
+
+GeometryPtr GeometryMerger::mergeGeometries(std::vector<GeometryPtr> geometries, const geos::geom::Envelope& envelope)
+{
+  LOG_DEBUG("Merging " << StringUtils::formatLargeNumber(geometries.size()) << " geometries...");
+
+  //  Create and start the threads
+  std::vector<std::thread> threads;
+  for (int i = 0; i < _maxThreads; ++i)
+    threads.push_back(thread(&GeometryMerger::mergeGeometryThread, this));
+
+  PolygonCompare compare(envelope);
+  //  Combine the geometries two at a time
+  _geometryCount = geometries.size();
+  while (geometries.size() > 1)
+  {
+    // Sort polygons using the Hilbert value. This increases the chances that nearby polygons will
+    // be merged early and speed up the union process.
+    sort(geometries.begin(), geometries.end(), compare);
+    LOG_TRACE("Remaining geometries: " << geometries.size());
+    //  Reserve the correct amount of space in the thread return vector
+    _geometryReturnMutex.lock();
+    _geometryReturn.resize(0);
+    _geometryReturn.reserve(geometries.size() / 2 + 1);
+    _geometryReturnMutex.unlock();
+    //  Create the stack of work
+    _geometryStackMutex.lock();
+    _geometryStack.empty();
+    for (size_t i = 0; i < geometries.size() - 1; i += 2)
+      _geometryStack.push(GeometryPair(geometries[i], geometries[i + 1]));
+    _geometryStackMutex.unlock();
+    //  Reset finished threads to zero each round of merges
+    _finishedThreads = 0;
+    //  Tell the threads to start working
+    _wait.notify_all();
```

Comment:
Using the `condition_variable` here allowed all of the threads to wait for the main thread to indicate that the work is ready for them by notifying them all at the same time.  Simplified the intra-thread communication process quite a bit.